### PR TITLE
bool inherited from number

### LINKED ISSUE #125003: bool as scalar in jit ir API not working as expected.
### 🐛 Describe the bug

in python code, we support bool literal as scalar, like this

``` python
a = torch.rand(1)
b = torch.clamp_max_(a, True)
```

but in `torch::jit::Graph` c++ API, if we use a bool node as input, then it can't find the candidate

you can test like this, in `test/cpp/jit/test_jit.cpp` 

``` cpp
TEST(IRTest, BoolMatchNumber) {
  Graph g;
  std::vector<Value*> inputs;
  auto b = g.addInput("max");
  b->setType(BoolType::get());
  inputs.push_back(b);
  Node* n = g.create(Symbol::fromQualString("aten::clamp_max_"), inputs, 1);
  ASSERT_EQ(n->schema().operator_name().name, "aten::clamp_max_");
}
```

the root cause is we define `BoolType` like this  `struct TORCH_API BoolType : public Type` 
which does not inherit from `NumberType`

### Versions

➜ python torch/utils/collect_env.py 
Collecting environment information...
PyTorch version: 2.2.0a0+git48a1237
Is debug build: True
CUDA used to build PyTorch: None
ROCM used to build PyTorch: N/A

OS: Ubuntu 22.04 LTS (x86_64)
GCC version: (Ubuntu 11.4.0-1ubuntu1~22.04) 11.4.0
Clang version: 14.0.5 (ssh://gerrit.habana-labs.com:29418/tpc_llvm10 ab097983d16bb8cedaa85e13d6123696de35047a)
CMake version: version 3.28.1
Libc version: glibc-2.35

Python version: 3.10.12 (main, Nov 20 2023, 15:14:05) [GCC 11.4.0] (64-bit runtime)
Python platform: Linux-5.15.0-102-generic-x86_64-with-glibc2.35
Is CUDA available: False
CUDA runtime version: No CUDA
CUDA_MODULE_LOADING set to: N/A
GPU models and configuration: No CUDA
Nvidia driver version: No CUDA
cuDNN version: No CUDA
HIP runtime version: N/A
MIOpen runtime version: N/A
Is XNNPACK available: True

CPU:
Architecture:                       x86_64
CPU op-mode(s):                     32-bit, 64-bit
Address sizes:                      43 bits physical, 48 bits virtual
Byte Order:                         Little Endian
CPU(s):                             12
On-line CPU(s) list:                0-11
Vendor ID:                          GenuineIntel
Model name:                         Intel(R) Xeon(R) Gold 5220R CPU @ 2.20GHz
CPU family:                         6
Model:                              85
Thread(s) per core:                 1
Core(s) per socket:                 6
Socket(s):                          2
Stepping:                           0
BogoMIPS:                           4389.68
Flags:                              fpu vme de pse tsc msr pae mce cx8 apic sep mtrr pge mca cmov pat pse36 clflush mmx fxsr sse sse2 ss ht syscall nx pdpe1gb rdtscp lm constant_tsc arch_perfmon nopl xtopology tsc_reliable nonstop_tsc cpuid tsc_known_freq pni pclmulqdq vmx ssse3 fma cx16 pcid sse4_1 sse4_2 x2apic movbe popcnt tsc_deadline_timer aes xsave avx f16c rdrand hypervisor lahf_lm abm 3dnowprefetch invpcid_single pti ssbd ibrs ibpb stibp tpr_shadow vnmi ept vpid ept_ad fsgsbase tsc_adjust bmi1 avx2 smep bmi2 invpcid avx512f avx512dq rdseed adx smap clflushopt clwb avx512cd avx512bw avx512vl xsaveopt xsavec xsaves arat pku ospke md_clear flush_l1d arch_capabilities
Virtualization:                     VT-x
Hypervisor vendor:                  VMware
Virtualization type:                full
L1d cache:                          384 KiB (12 instances)
L1i cache:                          384 KiB (12 instances)
L2 cache:                           12 MiB (12 instances)
L3 cache:                           71.5 MiB (2 instances)
NUMA node(s):                       1
NUMA node0 CPU(s):                  0-11
Vulnerability Gather data sampling: Unknown: Dependent on hypervisor status
Vulnerability Itlb multihit:        KVM: Mitigation: VMX disabled
Vulnerability L1tf:                 Mitigation; PTE Inversion; VMX flush not necessary, SMT disabled
Vulnerability Mds:                  Mitigation; Clear CPU buffers; SMT Host state unknown
Vulnerability Meltdown:             Mitigation; PTI
Vulnerability Mmio stale data:      Mitigation; Clear CPU buffers; SMT Host state unknown
Vulnerability Retbleed:             Mitigation; IBRS
Vulnerability Spec rstack overflow: Not affected
Vulnerability Spec store bypass:    Mitigation; Speculative Store Bypass disabled via prctl and seccomp
Vulnerability Spectre v1:           Mitigation; usercopy/swapgs barriers and __user pointer sanitization
Vulnerability Spectre v2:           Mitigation; IBRS, IBPB conditional, STIBP disabled, RSB filling, PBRSB-eIBRS Not affected
Vulnerability Srbds:                Not affected
Vulnerability Tsx async abort:      Not affected

Versions of relevant libraries:
[pip3] flake8==7.0.0
[pip3] habana-torch-dataloader==1.16.0+git2cfd41701
[pip3] habana-torch-plugin==1.16.0+git2cfd41701
[pip3] intel-extension-for-pytorch==2.2.0
[pip3] numpy==1.23.5
[pip3] pytorch-lightning==2.2.0.post0
[pip3] torch==2.2.0a0+git7e11024
[pip3] torch-debug==2.2.0a0+git7e11024
[pip3] torch_tb_profiler==0.4.0
[pip3] torchaudio==2.2.0+08901ad
[pip3] torchdata==0.7.1+5e6f7b7
[pip3] torchmetrics==1.3.2
[pip3] torchtext==0.17.0+400da5c
[pip3] torchvision==0.17.0+b2383d4
[pip3] triton==2.2.0
[conda] Could not collect

### PR DESCRIPTION
**Pull request by majian4work**
Fixes #125003 

cc @EikanWang @jgong5 @wenzhe-nrv @sanchitintel @gujinghui @PenghuiCheng @XiaobingSuper @jianyuh @mingfeima @ashokei @jingxu10 @min-jean-cho @yanbing-j @Guobing-Chen @Xia-Weiwen @snadampal

### CODE REVIEW

**Comment by colesbury** [comment]
@EikanWang - would you please review this PR?

**Comment by EikanWang** [comment]
LGTM.

**Review by EikanWang — APPROVED** [review]

**Review by EikanWang — CHANGES_REQUESTED** [review]
Please fix UT failures.

**Review by jgong5 — APPROVED** [review]

**Comment by Ma-Jian1** [comment]
@ezyang Could you help to take a review?

**Comment by ezyang** [comment]
You are still failing a bunch of tests

**Comment by Ma-Jian1** [comment]
hi all,
`torch/optim/asgd.py` is not part of this patch. what should I do next?
@EikanWang @ezyang

**Comment by ezyang** [comment]
I am not convinced I want to deal with the fallout from this in TorchScript. @davidberard98 for a second opinion

**Comment by davidberard98** [comment]
I'm seeing internal failures - not sure if I'll be able to get to this for a few days, but I'll need to take a closer look to understand what's failing. I'll report back once I have a better understanding.

**Comment by Ma-Jian1** [comment]
hi @davidberard98 ，does it still failed on your test？

**Comment by davidberard98** [comment]
@Ma-Jian1 yes - do you think you can check using torchvision to better understand the failures?

**Comment by Ma-Jian1** [comment]
hi @davidberard98 ，I clone torchvision into local pytorch fold，and run `python setup.py develop` in torchvision subfolder, and then ` python -m pytest -sv test/test_transforms_tensor.py -k test_convert_image_dtype`, I can reproduce 3 failed cases.
but after I reverted my patch, the failure still existed. 
seems not related to my patch?

investigated further:
I found the max_int literal `2147483647.0` will be captured as `2147483648.` in torch.script, so weird.

``` cpp
test/test_transforms_tensor.py::test_convert_image_dtype[out_dtype0-in_dtype0-cpu] graph(%image.1 : Tensor,
      %dtype.1 : int):
  %382 : int = prim::Constant[value=4294967294]()
  %377 : float = prim::Constant[value=9.2233720368547758e+18]()
  %376 : float = prim::Constant[value=2147483648.]()
  %333 : Function = prim::Constant[name="_max_value"]()
  %332 : str = prim::Constant[value="builtins.RuntimeError"]() # /home/jianma/repo/pytorch/vision/torchvision/transforms/_functional_tensor.py:79:18
  %329 : str = prim::Constant[value="The cast from {} to {} cannot be performed safely."]() # /home/jianma/repo/pytorch/vision/torchvision/transforms/_functional_tensor.py:78:18
  %321 : int = prim::Constant[value=7]() # /home/jianma/repo/pytorch/vision/torchvision/transforms/_functional_tensor.py:76:27
  %319 : bool = prim::Constant[value=1]() # /home/jianma/repo/pytorch/vision/torchvision/transforms/_functional_tensor.py:75:12
  %312 : int = prim::Constant[value=4]() # /home/jianma/repo/pytorch/vision/torchvision/transforms/_functional_tensor.py:75:68
  %311 : int = prim::Constant[value=3]() # /home/jianma/repo/pytorch/vision/torchvision/transforms/_functional_tensor.py:75:55
  %307 : int = prim::Constant[value=6]() # /home/jianma/repo/pytorch/vision/torchvision/transforms/_functional_tensor.py:75:27
  %206 : bool = prim::Constant[value=0]()
  %205 : NoneType = prim::Constant()
  %12 : int = prim::Constant[value=0]() # /home/jianma/repo/pytorch/vision/torchvision/transforms/_functional_tensor.py:71:24
  %64 : float = prim::Constant[value=0.001]() # /home/jianma/repo/pytorch/vision/torchvision/transforms/_functional_tensor.py:86:14
  %72 : float = prim::Constant[value=1.]() # /home/jianma/repo/pytorch/vision/torchvision/transforms/_functional_tensor.py:88:37
  %3 : int = prim::dtype(%image.1)
  %5 : bool = aten::eq(%3, %dtype.1) # /home/jianma/repo/pytorch/vision/torchvision/transforms/_functional_tensor.py:65:7
  %305 : Tensor = prim::If(%5) # /home/jianma/repo/pytorch/vision/torchvision/transforms/_functional_tensor.py:65:4
    block0():
      -> (%image.1)
    block1():
      %198 : bool = aten::is_floating_point(%image.1) # /home/jianma/repo/pytorch/vision/torchvision/transforms/_functional_tensor.py:68:7
      %201 : Tensor = prim::If(%198) # /home/jianma/repo/pytorch/vision/torchvision/transforms/_functional_tensor.py:68:4
        block0():
          %207 : Tensor = aten::tensor(%12, %dtype.1, %205, %206) # /home/jianma/repo/pytorch/vision/torchvision/transforms/_functional_tensor.py:71:11
          %208 : bool = aten::is_floating_point(%207) # /home/jianma/repo/pytorch/vision/torchvision/transforms/_functional_tensor.py:71:11
          %345 : Tensor = prim::If(%208) # /home/jianma/repo/pytorch/vision/torchvision/transforms/_functional_tensor.py:71:8
            block0():
              %215 : Tensor = aten::to(%image.1, %dtype.1, %206, %206, %205) # /home/jianma/repo/pytorch/vision/torchvision/transforms/_functional_tensor.py:72:19
              -> (%215)
            block1():
              %306 : int = prim::dtype(%image.1)
              %308 : bool = aten::eq(%306, %307) # /home/jianma/repo/pytorch/vision/torchvision/transforms/_functional_tensor.py:75:12
              %310 : bool = prim::If(%308) # /home/jianma/repo/pytorch/vision/torchvision/transforms/_functional_tensor.py:75:12
                block0():
                  %314 : int[] = prim::ListConstruct(%311, %312)
                  %315 : bool = aten::__contains__(%314, %dtype.1) # /home/jianma/repo/pytorch/vision/torchvision/transforms/_functional_tensor.py:75:45
                  -> (%315)
                block1():
                  -> (%206)
              %318 : bool = prim::If(%310) # /home/jianma/repo/pytorch/vision/torchvision/transforms/_functional_tensor.py:75:12
                block0():
                  -> (%319)
                block1():
                  %320 : int = prim::dtype(%image.1)
                  %322 : bool = aten::eq(%320, %321) # /home/jianma/repo/pytorch/vision/torchvision/transforms/_functional_tensor.py:76:12
                  %324 : bool = prim::If(%322) # /home/jianma/repo/pytorch/vision/torchvision/transforms/_functional_tensor.py:76:12
                    block0():
                      %326 : bool = aten::eq(%dtype.1, %312) # /home/jianma/repo/pytorch/vision/torchvision/transforms/_functional_tensor.py:76:45
                      -> (%326)
                    block1():
                      -> (%206)
                  -> (%324)
               = prim::If(%318) # /home/jianma/repo/pytorch/vision/torchvision/transforms/_functional_tensor.py:75:8
                block0():
                  %330 : int = prim::dtype(%image.1)
                  %msg.7 : str = aten::format(%329, %330, %dtype.1) # /home/jianma/repo/pytorch/vision/torchvision/transforms/_functional_tensor.py:78:18
                   = prim::RaiseException(%msg.7, %332) # /home/jianma/repo/pytorch/vision/torchvision/transforms/_functional_tensor.py:79:12
                  -> ()
                block1():
                  -> ()
              %334 : int = prim::CallFunction(%333, %dtype.1) # /home/jianma/repo/pytorch/vision/torchvision/transforms/_functional_tensor.py:87:24
              %max_val.7 : float = aten::Float(%334) # /home/jianma/repo/pytorch/vision/torchvision/transforms/_functional_tensor.py:87:18
              %337 : float = aten::add(%max_val.7, %72) # /home/jianma/repo/pytorch/vision/torchvision/transforms/_functional_tensor.py:88:27
              %338 : float = aten::sub(%337, %64) # /home/jianma/repo/pytorch/vision/torchvision/transforms/_functional_tensor.py:88:27
              %result.7 : Tensor = aten::mul(%image.1, %338) # /home/jianma/repo/pytorch/vision/torchvision/transforms/_functional_tensor.py:88:17
              %343 : Tensor = aten::to(%result.7, %dtype.1, %206, %206, %205) # /home/jianma/repo/pytorch/vision/torchvision/transforms/_functional_tensor.py:89:15
              -> (%343)
          -> (%345)
        block1():
           = prim::Print(%376) # /home/jianma/repo/pytorch/vision/torchvision/transforms/_functional_tensor.py:93:8
          %263 : Tensor = aten::tensor(%12, %dtype.1, %205, %206) # /home/jianma/repo/pytorch/vision/torchvision/transforms/_functional_tensor.py:97:11
          %264 : bool = aten::is_floating_point(%263) # /home/jianma/repo/pytorch/vision/torchvision/transforms/_functional_tensor.py:97:11
          %375 : Tensor = prim::If(%264) # /home/jianma/repo/pytorch/vision/torchvision/transforms/_functional_tensor.py:97:8
            block0():
              %image.65 : Tensor = aten::to(%image.1, %dtype.1, %206, %206, %205) # /home/jianma/repo/pytorch/vision/torchvision/transforms/_functional_tensor.py:98:20
              %273 : Tensor = aten::div(%image.65, %376) # /home/jianma/repo/pytorch/vision/torchvision/transforms/_functional_tensor.py:99:19
              -> (%273)
            block1():
               = prim::Print(%377) # /home/jianma/repo/pytorch/vision/torchvision/transforms/_functional_tensor.py:103:8
               = prim::Print(%377, %376, %382) # /home/jianma/repo/pytorch/vision/torchvision/transforms/_functional_tensor.py:116:12
              %image.91 : Tensor = aten::to(%image.1, %dtype.1, %206, %206, %205) # /home/jianma/repo/pytorch/vision/torchvision/transforms/_functional_tensor.py:117:20
              %373 : Tensor = aten::mul(%image.91, %382) # /home/jianma/repo/pytorch/vision/torchvision/transforms/_functional_tensor.py:118:19
              -> (%373)
          -> (%375)
      -> (%201)
  return (%305)
```

**Comment by Ma-Jian1** [comment]
I have to run "pip uninstall torch" **several times** to clean the virtual env.
now verified that the test passed without this patch.

**Comment by Ma-Jian1** [comment]
hi @davidberard98, the failed case in torchvision is related to cast int to float, which has an accuracy issue.
fixed it and add an extra unit test.

**Comment by Ma-Jian1** [comment]
hi @davidberard98 , 
could you help to take a review?
I have no mac device, and it shouldn't affect this unit test.

**Comment by Ma-Jian1** [comment]
@ezyang @davidberard98 Could you help to resolve the “Meta Internal-Only Changes Check”

**Review by EikanWang — APPROVED** [review]

**Comment by zhejiangxiaomai** [comment]
Hi, @davidberard98. We found this PR to be quite important during testing. Can we rebase and merge it in? me and @majian4work  can do rebase work.

**Comment by zhejiangxiaomai** [comment]
Hi @davidberard98 Any suggestion?

**Comment by zhejiangxiaomai** [comment]
Hi @davidberard98, can you please take a look at my comment?
